openpathsampling/beta/hooks.py

```diff
@@ -34,37 +41,95 @@ def after_simulation(self, sim, hook_state):
 
 
 class StorageHook(PathSimulatorHook):
-    """Standard hook for storage.
+    """
+    Standard hook for storage.
+
+    NOTE: Arguments passed to init take precedence over the corresponding
+          parameters of the PathSimulator this hook is attached to. They can
+          only be accessed through this hook, e.g. as hook.live_visualizer.
+
+    Parameters
+    ----------
+    storage : :class:`.Storage
+              where to save to; default ``None`` uses the simulation's
+              ``storage``
+    frequency : int
+                save frequency measured in steps; default ``None`` uses the
+                simulation's value for ``save_frequency``
+
     """
     implemented_for = ['before_simulation', 'after_step',
                        'after_simulation']
+
     def __init__(self, storage=None, frequency=None):
         self.storage = storage
         self.frequency = frequency
+        self._simulation = None
+
+    @property
+    def frequency(self):
+        if self._frequency is not None:
+            return self._frequency
+        elif self._simulation is not None:
+            return self._simulation.save_frequency
+        else:
+            raise SimulationNotFoundError("'frequency' has not "
+                                          + "been set and no hosting "
+                                          + "simulation known to get a value."
+                                          )
+
+    @frequency.setter
+    def frequency(self, val):
+        self._frequency = val
+
+    @property
+    def storage(self):
+        if self._storage is not None:
+            return self._storage
+        elif self._simulation is not None:
+            return self._simulation.storage
+        else:
+            raise SimulationNotFoundError("'storage' has not "
+                                          + "been set and no hosting "
+                                          + "simulation known to get a value."
+                                          )
+
+    @storage.setter
+    def storage(self, val):
+        self._storage = val
 
     def before_simulation(self, sim, **kwargs):
-        if self.storage is None:
-            self.storage = sim.storage
-        if self.frequency is None:
-            self.frequency = sim.save_frequency
+        self._simulation = sim
 
     def after_step(self, sim, step_number, step_info, state, results,
                    hook_state):
         if self.storage is not None:
             self.storage.save(results)
             if step_number % self.frequency == 0:
+                if sim.sample_set is not None:
+                    # some PathSimulators never set their sample_set
+                    # but PathSimulator.__init__ sets it to None
+                    sim.sample_set.sanity_check()
                 self.storage.sync_all()
 
     def after_simulation(self, sim, hook_state):
```

Comment:
This is something I am unsure about in both versions. In master we check for a storage not to be `None` that we then do not sync (which is strange, except if the two storages are always the same) and my version feels quite cumbersome (but I wasnt sure if we can expect the two storages to be the same).

Another possibility I thought about is the following:
```python
if self.storage is not None:
    self.storage.sync_all()
```
Given the fact that we set the `simulation.storage` as `self.storage` this should also work (assuming that the hook will not be transplanted to another simulation after calling `before_simulation`).

Reply:
The methods names on storage are really confusing, and don't do what you would expect from their names. (This is somewhat fixed in the new storage.)

In SimStore (new storage), the following is true, and I think the behavior is the same in netcdfplus:

* `save` is a function called `stash` in SimStore. It basically puts data into a staging area, but does not actually save it to disk
* `sync_all` dumps everything from the staging area to disk (and will probably be renamed in OPS 2.0 to something like `save_stashed`.) This is the part that actually takes significant time.

So you *always* want to put the data in the staging area (don't lose anything!) but you might only infrequently want to actually dump that to disk (large fraction of cost for toy models is saving to disk, so batch it together).

Actually, looking a little more carefully at this, you'll probably want to add something based on the try/except clause that has been added to `save_current_step`:

https://github.com/openpathsampling/openpathsampling/blob/b6da5ddf14981c6624c2607f3b00ce0ec120c85d/openpathsampling/pathsimulators/path_sampling.py#L115-L120

(where `storage.steps.save` can be replaced by `storage.save` to save any general object.)